### PR TITLE
Add permissions check to valid_owner

### LINKED ISSUE #40: Owners check does not support child teams with inherited repo perms
**Description**

When running the "owners" check on a child GitHub team which has inherited its repo perms from a parent, the GitHub action reports: 
>Team "engineering-foo" does not exist in organization "some-org" or has no permissions associated with the repository.

I took a look through the code, and I think the issue comes down to the generation of the teams list [here](https://github.com/mszostok/codeowners-validator/blob/d18c45dcfcc4208fb3f50f33631f752e1fd896aa/internal/check/valid_owner.go#L104). That appears to be making a call to the [`/repos/{owner}/{repo}/teams`](https://docs.github.com/en/rest/reference/repos#list-repository-teams) GitHub API endpoint, which, as far as I can tell, does not return teams that have inherited their repo permissions from another team.

This is for version 4.0 of this repo.

**Expected result**

This is a team that has access to the repo (indirectly), so I'd expect it the "owners" check to pass.

**Actual result**

The "owners" check fails with the error message above.

**Steps to reproduce**

I've created an example PR here: https://github.com/test-verify-codeowners/owners-bug/pull/1.

For posterity, to setup that org, I:
1. Created it, added a repo, and added the `parent-team` team
2. Created `child-team-inherit` which just inherits `parent-team`'s perms, and a `child-team-override`, where I overwrote the inherited perms to explicitly give it access (described in the workaround section below.)
3. Added the appropriate personal access token to the orgs' secrets.
4. Made the changes visible in that PR and opened the PR (more info in that PR's description).

**Troubleshooting**

Curling the repos endpoint with `curl -H "Authorization: token <token>" https://api.github.com/repos/test-verify-codeowners/owners-bug/teams` gives us:

```json
[
  {
    "name": "parent-team",
    "id": 3934999,
    "node_id": "MDQ6VGVhbTM5MzQ5OTk=",
    "slug": "parent-team",
    "description": "",
    "privacy": "closed",
    "url": "https://api.github.com/organizations/68017749/team/3934999",
    "html_url": "https://github.com/orgs/test-verify-codeowners/teams/parent-team",
    "members_url": "https://api.github.com/organizations/68017749/team/3934999/members{/member}",
    "repositories_url": "https://api.github.com/organizations/68017749/team/3934999/repos",
    "permission": "pull",
    "parent": null
  },
  {
    "name": "child-team-override",
    "id": 3935004,
    "node_id": "MDQ6VGVhbTM5MzUwMDQ=",
    "slug": "child-team-override",
    "description": "This one also inherits, but also overrides.",
    "privacy": "closed",
    "url": "https://api.github.com/organizations/68017749/team/3935004",
    "html_url": "https://github.com/orgs/test-verify-codeowners/teams/child-team-override",
    "members_url": "https://api.github.com/organizations/68017749/team/3935004/members{/member}",
    "repositories_url": "https://api.github.com/organizations/68017749/team/3935004/repos",
    "permission": "pull",
    "parent": {
      "name": "parent-team",
      "id": 3934999,
      "node_id": "MDQ6VGVhbTM5MzQ5OTk=",
      "slug": "parent-team",
      "description": "",
      "privacy": "closed",
      "url": "https://api.github.com/organizations/68017749/team/3934999",
      "html_url": "https://github.com/orgs/test-verify-codeowners/teams/parent-team",
      "members_url": "https://api.github.com/organizations/68017749/team/3934999/members{/member}",
      "repositories_url": "https://api.github.com/organizations/68017749/team/3934999/repos",
      "permission": "pull"
    }
  }
]
```

Note that [`child-team-override`'s repo list](https://github.com/orgs/test-verify-codeowners/teams/child-team-override/repositories) is visually the same as [`child-team-inherit`'s repo list](https://github.com/orgs/test-verify-codeowners/teams/child-team-inherit/repositories), but only `parent-team` and `child-team-override` show up in the above response. 

**Workarounds**

We've found a workaround for this, which is to manually give the child teams perms on the repo in question.

1. Go to the team's repository settings (e.g. https://github.com/orgs/test-verify-codeowners/teams/child-team-override/repositories.

2. Find the repo in question

3. Open its permissions dropdown and select the level of perms you want to give it (which should already be selected).

    image

It will look basically the same, but it you cURL the endpoint again, it should now show up in the response, and the action will work.

**Potential Solutions**

I'm not really sure if this counts as a bug on your side. I haven't found any references to this behavior in the GitHub API docs, it seems strange, so it could be a bug on their end. I'm planning to reach out to GitHub support today and see what they say.

However, I thought I'd open this and see what you think (and in case others run into this issue.) If this behavior is intentional on their end (or just unlikely to change in the near future), I wonder if it would be possible to substitute calls to other endpoints. For example, hitting `https://api.github.com/orgs/test-verify-codeowners/teams` gives us a repositories url of `https://api.github.com/organizations/68017749/team/3935001/repos` for `child-team-inherit`, which does return the repo.

Also, thank you! This project is very helpful for us.

### PR DESCRIPTION
**Pull request by highb**
**Description**

Changes proposed in this pull request:

- Changes the initial check for the validity/existence of a team to use the Teams List Teams API endpoint
- Adds a permissions check that uses the Teams IsTeamRepoBySlug endpoint, to check that the owner has write permissions on the repo

I haven't contributed to many Go Projects, so hopefully I'm following good conventions.

**Related issue(s)**

Fixes #40

### CODE REVIEW

**Comment by highb** [comment]
I'm not surprised to see that I should add more unit tests. I'll work on that when I have a moment.

**Comment by jspiro** [comment]
This is just what we need, thanks for fixing this!

**Comment by jspiro** [comment]
I don't see an easy way to increase coverage in this file. All of the active check code is uncovered as it hits GitHub APIs – I can't tell if they're mocked nor how, but when I gave it a trivial try, the github API segfaulted (!).

This might explain the file being 50% coverage target. @mszostok what's the path forward here? This is a critical fix for teams using parent team hierarchies, and the code looks good to me!

**Comment by highb** [comment]
@mszostok Thanks for taking a look! If there's anything I can do to help out, please ask.

File: internal/check/valid_owner.go
Comment (mszostok) [inline internal/check/valid_owner.go:218]:
```suggestion
		for k, v := range repo.GetPermissions() {
```

to ensure that we will not have panic here

File: internal/check/valid_owner.go
Comment (mszostok) [inline internal/check/valid_owner.go:232]: this case is not needed as there is no other logic after switch

File: internal/check/valid_owner.go
Comment (mszostok) [inline internal/check/valid_owner.go:240]:
```suggestion
		return newValidateError("Team %q cannot review PRs on %q as it doesn't have a write permission.", team, v.orgRepoName)
```
we can be more descriptive here and says that we failed on permission check

File: internal/check/valid_owner.go
Comment (mszostok) [inline internal/check/valid_owner.go:219]:
```suggestion
			if !v {
```
Unfortunately, the linter is yelling at us 😄 

https://travis-ci.com/github/mszostok/codeowners-validator/jobs/471190231

File: internal/check/valid_owner.go
Comment (mszostok) [inline internal/check/valid_owner.go:228]: I know that the `admin` and `push` are the valid ones, are you sure that the `maintain` and `triage` also allow approvals?

From what I can see `triage` seems to do not give us "write" perms
Screenshot 2021-01-20 at 21 27 41

File: internal/check/valid_owner.go
Comment (mszostok) [inline internal/check/valid_owner.go:199]: Thanks for adding that functionality, it was also requested some time ago (https://github.com/mszostok/codeowners-validator/issues/21)

Unfortunately, this approach increases the number of external call to GitHub, probably in the near feature it will be good to switch to GraphQL which will remove problem with over fetched data, and we will be able to merge the `ListTeams` and `IsTeamRepoBySlug` into single query,  see: https://github.com/mszostok/codeowners-validator/issues/21#issuecomment-716044156

but this is for future, for now, it lgtm 👍

**Review by mszostok — COMMENTED** [review]
Once again thanks for your contribution!

I executed integration tests, and it works properly as I also simulated that some teams have only "read" permission, and they were detected properly 👍 

I left a few comments which are minor. After resolving them we can merge the pull-request 🚀 

If you do not have time for that I can take care of it as they are just small adjustments - if necessary ping me under this pull request 🙂

File: internal/check/valid_owner.go
Comment (jspiro) [reply internal/check/valid_owner.go:232]: I think we were just being abundantly clear as to our intention, in case it was ever refactored. But addressed.

File: internal/check/valid_owner.go
Comment (jspiro) [reply internal/check/valid_owner.go:219]: Ah, interesting. The vendored golangci-lint installed with go on brew doesn't pick this up. Installing unvendored reproduced it.

File: internal/check/valid_owner.go
Comment (jspiro) [reply internal/check/valid_owner.go:228]: Good call, triage is incorrect.

File: internal/check/valid_owner.go
Comment (jspiro) [reply internal/check/valid_owner.go:199]: That is quite true. There are also potentially issues with GraphQL, which we've experienced with another project we contribute to, the GitHub Terraform provider – some calls because extremely, extremely slow in larger orgs (branch protection) and it annihilated the performance compared to REST. I doubt it'd be an issue here, just sharing our experience.

Comment (jspiro) [reply internal/check/valid_owner.go:199]: I'll add a todo.

File: internal/check/valid_owner.go
Comment (jspiro) [reply internal/check/valid_owner.go:240]: Done, cleaned up some extra long lines too.

File: internal/check/valid_owner.go
Comment (jspiro) [reply internal/check/valid_owner.go:218]: Done

**Comment by highb** [comment]
@mszostok FYI @jspiro is my rockstar colleague who I granted collaborator access on my fork to. 😄 It looks like he's essentially addressed everything and I'm 👍 on the changes.

**Comment by jspiro** [comment]
@mszostok As you indicated, Travis is hitting API limits, presumably due to unauthenticated requests having a limit of 60 per hour per IP. I suppose this is a good reason for GQL ;-)

```
        +    [err] line 8: Cannot initialize organization member list: GET https://api.github.com/orgs/gh-codeowners/members?per_page=100: 403 API rate limit exceeded for 207.254.16.35. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.) [rate reset in 35m37s]
253
```

Have you considered porting this to GitHub Actions? I suspect that would paper over this issue, at least.

**Comment by jspiro** [comment]
If this is acceptable and gets merged, please remember to tag it :-)

**Comment by mszostok** [comment]
> @mszostok As you indicated, Travis is hitting API limits, presumably due to unauthenticated requests having a limit of 60 per hour per IP. I suppose this is a good reason for GQL ;-)
> 
> ```
>         +    [err] line 8: Cannot initialize organization member list: GET https://api.github.com/orgs/gh-codeowners/members?per_page=100: 403 API rate limit exceeded for 207.254.16.35. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.) [rate reset in 35m37s]
> 253
> ```
> 
> Have you considered porting this to GitHub Actions? I suspect that would paper over this issue, at least.

Actually, it is already available as the GitHub Action https://github.com/mszostok/codeowners-validator/blob/master/docs/gh-action.md. With that, you "mount" the GitHub token which allows more API calls.

About porting that to GraphQL, I'm waiting for more reports if this is really an issue.

File: internal/check/valid_owner.go
Comment (mszostok) [reply internal/check/valid_owner.go:201]: thx for the heads-up 👍

**Comment by mszostok** [comment]
> @mszostok As you indicated, Travis is hitting API limits, presumably due to unauthenticated requests having a limit of 60 per hour per IP. I suppose this is a good reason for GQL ;-)
> 
> ```
>         +    [err] line 8: Cannot initialize organization member list: GET https://api.github.com/orgs/gh-codeowners/members?per_page=100: 403 API rate limit exceeded for 207.254.16.35. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.) [rate reset in 35m37s]
> 253
> ```
> 
> Have you considered porting this to GitHub Actions? I suspect that would paper over this issue, at least.

Yes, this is on my TODO list, as for all other project I already use the GitHub Actions. Maybe during this weekend I will have some time to do it and add also release pipeline cause right now it is pain in the ass as it is manual 😄

**Comment by mszostok** [comment]
@jspiro & @highb  Thanks for your contribution! LGTM 🚀

**Review by mszostok — APPROVED** [review]
@jspiro & @highb  Thanks for your contribution! LGTM 🚀

**Comment by highb** [comment]
:shipit: Looking forward to rolling this out on all our repos! Thanks for putting the project out there!

**Pull request merged**
